File: Sources/Scratchpad/Drawing/StrokeGeometry.swift
import CoreGraphics

/// Pure point-to-polyline distance math, used to hit-test the eraser
/// against existing strokes. Kept free of AppKit so it stays testable and
/// portable, like `StrokeSmoothing`.
enum StrokeGeometry {
    static func minimumDistance(from point: CGPoint, toPolylineThrough points: [StrokePoint]) -> Double {
        guard let first = points.first else { return .infinity }
        guard points.count > 1 else { return distance(point, first.cgPoint) }

        var closest = Double.infinity
        for index in 0..<(points.count - 1) {
            let segmentDistance = distance(point, toSegmentFrom: points[index].cgPoint, to: points[index + 1].cgPoint)
            closest = min(closest, segmentDistance)
        }
        return closest
    }

    private static func distance(_ a: CGPoint, _ b: CGPoint) -> Double {
        Double(hypot(a.x - b.x, a.y - b.y))
    }

    private static func distance(_ point: CGPoint, toSegmentFrom a: CGPoint, to b: CGPoint) -> Double {
        let dx = b.x - a.x
        let dy = b.y - a.y
        let lengthSquared = dx * dx + dy * dy
        guard lengthSquared > 0 else { return distance(point, a) }

        let t = max(0, min(1, ((point.x - a.x) * dx + (point.y - a.y) * dy) / lengthSquared))
        let projection = CGPoint(x: a.x + t * dx, y: a.y + t * dy)
        return distance(point, projection)
    }
}

File: Windows/Scratchpad/Drawing/StrokeGeometry.cs
namespace Scratchpad.Drawing;

/// Pure point-to-polyline distance math, used to hit-test the eraser against
/// existing strokes. Kept free of WPF so it stays testable and portable,
/// like `StrokeSmoothing`. Identical algorithm to the macOS app's
/// `StrokeGeometry`.
public static class StrokeGeometry
{
    public static double MinimumDistance(StrokePoint point, IReadOnlyList<StrokePoint> polyline)
    {
        if (polyline.Count == 0) return double.PositiveInfinity;
        if (polyline.Count == 1) return Distance(point, polyline[0]);

        var closest = double.PositiveInfinity;
        for (var index = 0; index < polyline.Count - 1; index++)
        {
            var segmentDistance = DistanceToSegment(point, polyline[index], polyline[index + 1]);
            closest = Math.Min(closest, segmentDistance);
        }
        return closest;
    }

    private static double Distance(StrokePoint a, StrokePoint b) =>
        Math.Sqrt(Math.Pow(a.X - b.X, 2) + Math.Pow(a.Y - b.Y, 2));

    private static double DistanceToSegment(StrokePoint point, StrokePoint a, StrokePoint b)
    {
        var dx = b.X - a.X;
        var dy = b.Y - a.Y;
        var lengthSquared = dx * dx + dy * dy;
        if (lengthSquared <= 0) return Distance(point, a);

        var t = Math.Max(0, Math.Min(1, ((point.X - a.X) * dx + (point.Y - a.Y) * dy) / lengthSquared));
        var projection = new StrokePoint(a.X + t * dx, a.Y + t * dy);
        return Distance(point, projection);
    }
}
